### PR TITLE
[dev] Improving logging of blocking packages.

### PR DESCRIPTION
**Pull request by PawelWMS**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [X] The toolchain has been rebuilt successfully (or no changes were made to it)
- [X] The toolchain/worker package manifests are up-to-date
- [X] Any updated packages successfully build (or no packages were changed)
- [X] All package sources are available
- [X] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`)
- [X] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [X] All source files have up-to-date hashes in the `*.signatures.json` files
- [X] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [X] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

The change improves how blocking packages are displayed in the logs to include the whole dependency chain, which lead to the conclusion that a particular package is blocking a given SRPM.

Before:
```
level=info msg="136: at-spi2-atk-2.34.2-1.cm1.src.rpm - 2 total unmet dependencies"
level=debug msg="--> gtk3-3.24.28-3.cm1.src.rpm"
level=debug msg="--> pkgconfig(atk-bridge-2.0)"
```

After:
```
level=info msg="136: at-spi2-atk-2.34.2-1.cm1.src.rpm - 1 total unmet dependencies"
level=debug msg="--> pkgconfig(atk-bridge-2.0): gtk2-devel-2.24.32-8.cm1.x86_64.rpm [gtk2-2.24.32-8.cm1.src.rpm] -> gtk2-2.24.32-8.cm1.x86_64.rpm [gtk2-2.24.32-8.cm1.src.rpm] -> gtk-update-icon-cache-3.24.28-3.cm1.x86_64.rpm [gtk3-3.24.28-3.cm1.src.rpm] -> pkgconfig(atk-bridge-2.0)"
```

Note that after the changes there's only one unmet dependency as `gtk3-3.24.28-3.cm1.src.rpm` from the original was actually blocked itself by unresolved `pkgconfig(atk-bridge-2.0)`, so we only keep the "final" blocker.

The change also introduces a new `StateBuildError` node state to distinguish between build nodes, which had attempted to build because they had their dependencies satisfied but failed, and build nodes, which were blocked from building due to missing dependencies.

###### Change Log  

- Changed logging of indirect blocking packages.
- Introduces a new `StateBuildError` node state.
- Modified the `sliceutils.Find` function to work with slices of arbitrary types.

###### Does this affect the toolchain?  

No.

###### Test Methodology

- Local tool runs against graph .dot files from our builds.
- Package build in the pipelines.

### CODE REVIEW

File: toolkit/tools/internal/sliceutils/sliceutils.go [hunk 1]
```diff
@@ -26,3 +31,8 @@ func FindMatches(slice []string, isMatch func(string) bool) []string {
 	}
 	return result
 }
+
+// StringMatch is intended to be used with "Find" for slices of strings.
+func StringMatch(expected, given interface{}) bool {
```
Comment (mfrw) [inline toolkit/tools/internal/sliceutils/sliceutils.go:36]: I am of the opinion
```suggestion
// DoStringsMatch is intended to be used with "Find" for slices of strings.
func DoStringsMatch(expected, given interface{}) bool {
```

File: toolkit/tools/pkgworker/pkgworker.go [hunk 2]
```diff
@@ -309,7 +309,7 @@ func removeLibArchivesFromSystem() (err error) {
 
 		// Skip directories that are meant for device files and kernel virtual filesystems.
 		// These will not contain .la files and are mounted into the safechroot from the host.
-		if info.IsDir() && sliceutils.Find(dirsToExclude, path) != sliceutils.NotFound {
+		if info.IsDir() && sliceutils.Find(dirsToExclude, path, sliceutils.StringMatch) != sliceutils.NotFound {
```
Comment (mfrw) [inline toolkit/tools/pkgworker/pkgworker.go:312]:
```suggestion
		if info.IsDir() && sliceutils.Find(dirsToExclude, path, sliceutils.DoStringsMatch) != sliceutils.NotFound {
```

File: toolkit/tools/scheduler/schedulerutils/preparerequest.go [hunk 3]
```diff
@@ -66,7 +66,7 @@ func canUseCacheForNode(pkgGraph *pkggraph.PkgGraph, node *pkggraph.PkgNode, pac
 
 	// Check if the node corresponds to an entry in packagesToRebuild
 	specName := strings.TrimSuffix(filepath.Base(node.SpecPath), specSuffix)
-	canUseCache = (sliceutils.Find(packagesToRebuild, specName) == -1)
+	canUseCache = (sliceutils.Find(packagesToRebuild, specName, sliceutils.StringMatch) == -1)
```
Comment (mfrw) [inline toolkit/tools/scheduler/schedulerutils/preparerequest.go:69]:
```suggestion
	canUseCache = (sliceutils.Find(packagesToRebuild, specName, sliceutils.DoStringsMatch) == -1)
```

File: toolkit/tools/graphanalytics/graphanalytics.go [hunk 4]
```diff
@@ -275,7 +313,7 @@ func printReversedMap(data map[string][]string, valueDescription string, maxResu
 // printMap will sort and print the largest entries, using the valueDescription in the format.
 func printMap(data map[string][]string, valueDescription string, maxResults int) {
 	const inverse = false
-	pairList := sortMap(data, false)
+	pairList := sortMap(data, inverse)
```
[the remark this answers is not in the document]
Comment (PawelWMS) [reply toolkit/tools/graphanalytics/graphanalytics.go:316]: That's what we're trying to stick to when it comes to using consts that are not obvious. We've already done it in another place in this file and I think it's just been overlooked during some refactoring in the past.:)

File: toolkit/docs/how_it_works/3_package_building.md [hunk 5]
```diff
@@ -75,6 +75,8 @@ The graph contains several types of nodes, with various states. As the graph is
 >
 > `StateBuild`: Should be built
 >
+> `StateBuildError`: Package had all of its dependencies satisfied but failed to build for other reasons.
```
[the remark this answers is not in the document]
Comment (PawelWMS) [reply toolkit/docs/how_it_works/3_package_building.md:78]: I think this one is fine: I'm trying to say that the package has already failed to build at the time this status is set.

File: toolkit/tools/internal/sliceutils/sliceutils.go (hunk 1 above)
Comment (PawelWMS) [reply toolkit/tools/internal/sliceutils/sliceutils.go:36]: In general for functions I'm also in favour of using names that describe an action `doX`, but in this case I'm using a noun "match" on purpose to describe the type of match between the elements we're interested in for a particular call to `sliceutils.Find`. It's the same idea as the `graphpath.NullHeuristic` elsewhere in my changes (that ones' part of an external library, though).

File: toolkit/tools/scheduler/schedulerutils/preparerequest.go (hunk 3 above)
Comment (PawelWMS) [reply toolkit/tools/scheduler/schedulerutils/preparerequest.go:69]: Statying with the current one. See: previous comment.

File: toolkit/tools/internal/sliceutils/sliceutils.go [hunk 6]
```diff
@@ -3,16 +3,21 @@
 
 package sliceutils
 
+import "reflect"
+
 // NotFound value is returned by Find(), if a given value is not present in the slice.
 const NotFound = -1
 
 // Find returns an index of the first occurrence of thing in slice, or -1 if it does not appear in the slice.
```
Comment (jslobodzian) [inline toolkit/tools/internal/sliceutils/sliceutils.go:11]: >f thing in slice, or [](http://example.com/codeflow?start=50&length=20)

update comment

**Review by thcrain-msft — APPROVED** [review]

**Pull request merged**